src/librbd/io/ImageRequestWQ.cc

```diff
@@ -492,6 +531,110 @@ void ImageRequestWQ<I>::aio_compare_and_write(AioCompletion *c,
   trace.event("finish");
 }
 
+template <typename I>
+bool ImageRequestWQ<I>::block_overlapping_io(ImageExtents* in_flight_image_extents,
+                                             uint64_t off,
+                                             uint64_t len) {
+  CephContext *cct = m_image_ctx.cct;
+  ldout(cct, 20) << "ictx=" << &m_image_ctx
+                 << "off: " << off << " len: " << len <<dendl;
+
+  if(!len) {
+    return false;
+  }
+
+  if (!in_flight_image_extents->empty() && in_flight_image_extents->intersects(off, len)) {
+    return true;
+  }
+
+  in_flight_image_extents->insert(off, len);
+  return false;
+}
+
+template <typename I>
+void ImageRequestWQ<I>::unblock_flushes(uint64_t tid) {
+  CephContext *cct = m_image_ctx.cct;
+  ldout(cct, 20) << "ictx=" << &m_image_ctx << dendl;
+  ZTracer::Trace trace;
+
+  ordering_lock.lock();
+  auto io_tid_it = m_queued_or_blocked_io_tids.begin();
+  while (true) {
+    auto it = m_queued_flushes.begin();
+    if (it == m_queued_flushes.end() ||
+        (io_tid_it != m_queued_or_blocked_io_tids.end() &&
+        *io_tid_it < it->first)) {
+      break;
+    }
+
+    auto blocked_flush = *it;
+    ldout(cct, 20) << "unblocking flush: tid" << blocked_flush.first << dendl;
+    blocked_flush.second.c->init_time(util::get_image_ctx(&m_image_ctx), AIO_TYPE_FLUSH);
+    if (blocked_flush.second.native_async && m_image_ctx.event_socket.is_valid()) {
```

Comment:
Nit: why couldn't this be set up in `aio_flush`?

Reply:
Done